Collision filter

akaszynski opened the pull request:
This PR adds support for [vtkCollisionDetectionFilter](https://vtk.org/doc/nightly/html/classvtkCollisionDetectionFilter.html#af310f9312cd84f971a8a7a77f1f2d1eb). Includes a docstring example and full example.

![collision_movie](https://user-images.githubusercontent.com/11981631/126882222-407e7361-ada7-49f6-a7cf-1e72df67964b.gif)

MatthewFlamm commented on examples/01-filter/collisions.py:
```diff
@@ -0,0 +1,77 @@
+
+    colision_mask = np.zeros(sphere0.n_cells, dtype=bool)
+    if n_contacts:
+        colision_mask[col['ContactCells']] = True
```
```suggestion
        collision_mask[col['ContactCells']] = True
```

MatthewFlamm commented on examples/01-filter/collisions.py:
```diff
@@ -0,0 +1,77 @@
+    colision_mask = np.zeros(sphere0.n_cells, dtype=bool)
+    if n_contacts:
+        colision_mask[col['ContactCells']] = True
+    sphere0['collisions'] = colision_mask
```
```suggestion
    sphere0['collisions'] = collision_mask
```

MatthewFlamm commented on examples/02-plot/linked.py:
```diff
@@ -29,9 +29,8 @@
-p.show(auto_close=False)
 p.open_gif("linked.gif")
-
+ 
```
```suggestion

```

MatthewFlamm commented on pyvista/core/filters/poly_data.py:
```diff
@@ -2601,3 +2601,114 @@ def strip(poly_data, join=False, max_length=1000, pass_cell_data=False,
+
+        Parameters
+        ----------
+        other_mesh : pyvista.PolyData
```
Does it make sense to check that `isinstance(other_mesh, pyvista.PolyData)`?

akaszynski replied:
Yes, and out of convenience, I'll extract the surface if it's not polydata.

akaszynski replied:
Actually, according to [vtkCollisionDetectionFilter_8h_source.html](https://vtk.org/doc/nightly/html/vtkCollisionDetectionFilter_8h_source.html), I'll need to check if the meshes are triangular as well.

akaszynski replied:
Added a check and convert.

MatthewFlamm commented on pyvista/core/filters/poly_data.py:
```diff
@@ -2601,3 +2601,114 @@ def strip(poly_data, join=False, max_length=1000, pass_cell_data=False,
+            * 2 - Half contacts. Quickly find the first contact point.
+
+        box_tolerance : float, optional
+            OBB tolerance in world coordinates.
```
I don't know what OBB is.  I think it would be helpful to have the non-acronym somewhere.

akaszynski replied:
Sorry, copied this right out of the header.

```suggestion
             Oriented bounding box (OBB) tree tolerance in world coordinates.
```

MatthewFlamm commented on pyvista/core/filters/poly_data.py:
```diff
@@ -2601,3 +2601,114 @@ def strip(poly_data, join=False, max_length=1000, pass_cell_data=False,
+            attribute named ``"ContactCells"``.  Array only exists
+            when there are collisions.
+
+        :class:`int`
```
Is the class notation here necessary?

akaszynski replied:
It is to get it rendered correctly as a link (which isn't really necessary, but looks cool).  There's an issue with `sphinx.ext.napoleon` when `napoleon_use_rtype = False` to get intersphinx working.  Since we're not doing this anywhere else in our code, we can pass on this.

```suggestion
        int
```

MatthewFlamm commented on pyvista/core/filters/poly_data.py:
```diff
@@ -2601,3 +2601,114 @@ def strip(poly_data, join=False, max_length=1000, pass_cell_data=False,
+        alg.SetCollisionMode(contact_mode)
+        alg.SetGenerateScalars(generate_scalars)
+        _update_alg(alg, progress_bar, 'Computing collisions')
+        return pyvista.wrap(alg.GetOutput()), alg.GetNumberOfContacts()
```
Does it make sense to have a comment here to explain why `_get_output` isn't used for the first element of the returned tuple?

akaszynski replied:
I've just neglected to use `_get_output`.  Fixed.

MatthewFlamm commented on tests/test_filters.py:
```diff
@@ -1582,3 +1582,12 @@ def test_subdivide_adaptive(sphere, inplace):
+    output, n_collision = sphere.collision(moved_sphere)
+    assert isinstance(output, pyvista.PolyData)
+    assert n_collision > 40
+    assert 'ContactCells' in output.field_arrays
```
It would be nice to also test that non-intersecting meshes returns what is expected, but hardly necessary.

akaszynski replied:
Added.  It's a cheap test.

MatthewFlamm commented on pyvista/core/filters/poly_data.py:
```diff
@@ -2601,3 +2601,114 @@ def strip(poly_data, join=False, max_length=1000, pass_cell_data=False,
+            Number of cells in each OBB.
+
+        generate_scalars : bool, optional
+            Flag to visualize the contact cells.  If ``True``, the
```
This section gives me the impression that this only makes sense when supplying a transform to vtk directly that results in a collision.  Does this option actually result in differently colored messages in this implementation?

akaszynski replied:
> This section gives me the impression that this only makes sense when supplying a transform to vtk directly that results in a collision. Does this option actually result in differently colored messages in this implementation?

The mesh will contain colored "colliding" cells when this flag is True regardless of the transform supplied to VTK.

MatthewFlamm commented on pyvista/core/filters/poly_data.py:
```diff
@@ -2601,3 +2601,114 @@ def strip(poly_data, join=False, max_length=1000, pass_cell_data=False,
+              with two points per collision
+            * 1 - First contact. Find all the contacting cell pairs
+              with one point per collision.
+            * 2 - Half contacts. Quickly find the first contact point.
```
Is there such a thing as "first contact point" in this implementation?  Maybe I misunderstand exactly what multiple collisions means.

akaszynski replied:
> I think the descriptions for First and Half contacts are reversed here.

Quite right.

> Is there such a thing as "first contact point" in this implementation?

According to the header file and their documentation [vtkCollisionDetection](https://vtk.org/doc/nightly/html/classvtkCollisionDetectionFilter.html#details), yes.  As for the "how", I don't see any documentation on how it computes the collisions.

darikg commented:
I think `contact_mode` should be a string or enum, for the sake of more obvious readability

akaszynski commented on pyvista/core/filters/poly_data.py:
```diff
@@ -2601,3 +2601,114 @@ def strip(poly_data, join=False, max_length=1000, pass_cell_data=False,
+
+            * 0 - All contacts. Find all the contacting cell pairs
+              with two points per collision
+            * 1 - First contact. Find all the contacting cell pairs
+              with one point per collision.
+            * 2 - Half contacts. Quickly find the first contact point.
```
```suggestion
            * 1 - First contact. Quickly find the first contact point.
            * 2 - Half contacts. Find all the contacting cell pairs
              with one point per collision.
```

akaszynski commented on pyvista/core/filters/poly_data.py:
```diff
@@ -2601,3 +2601,114 @@ def strip(poly_data, join=False, max_length=1000, pass_cell_data=False,
+        >>> _ = pl.add_mesh(mesh_b, color='tan', line_width=5, opacity=0.7, 
+        ...                 show_edges=True)
+        >>> pl.show()
+
```
```suggestion
        Warnings
        --------
        Currently only triangles are processed. Use :func:`PolyDataFilters.triangulate`
        to convert any strips or polygons to triangles.

```

akaszynski commented on pyvista/core/filters/poly_data.py:
```diff
@@ -2601,3 +2601,114 @@ def strip(poly_data, join=False, max_length=1000, pass_cell_data=False,
+        alg.SetCollisionMode(contact_mode)
+        alg.SetGenerateScalars(generate_scalars)
+        _update_alg(alg, progress_bar, 'Computing collisions')
+        return pyvista.wrap(alg.GetOutput()), alg.GetNumberOfContacts()
```
Sorry, should have just used `_get_output`.

```suggestion
        return _get_output(alg), alg.GetNumberOfContacts()
```

MatthewFlamm commented:
I played with this to try and understand it myself, but came away more confused.  Maybe this is my problem, but there are some funny observations:

```py
import numpy as np
import pyvista as pv
sphere0 = pv.Sphere()
sphere0['collisions'] = np.zeros(sphere0.n_cells, dtype=bool)
sphere1 = pv.Sphere(radius=0.6, center=(-1, 0, 0))
col, n_contacts = sphere0.collision(sphere1)
print(n_contacts)
print(col)
print(col.array_names)
col, n_contacts = sphere0.collision(sphere1, generate_scalars=True)
print(n_contacts)
print(col)
print(col.array_names)
```

yields:

```txt
84
PolyData (0x7f49542ae5e0)
  N Cells:      1680
  N Points:     842
  X Bounds:     -4.993e-01, 4.993e-01
  Y Bounds:     -4.965e-01, 4.965e-01
  Z Bounds:     -5.000e-01, 5.000e-01
  N Arrays:     3

['collisions', 'ContactCells', 'Normals']
84
PolyData (0x7f4953959ca0)
  N Cells:      1680
  N Points:     842
  X Bounds:     -4.993e-01, 4.993e-01
  Y Bounds:     -4.965e-01, 4.965e-01
  Z Bounds:     -5.000e-01, 5.000e-01
  N Arrays:     3

['ContactCells', 'Normals']
```

Note that the `col` object the second time around using `generate_scalars=True` claims to have 3 arrays, but only 2 are available, where `collision` has dissapeared.  The other weird observation is that if I delete the line `sphere0['collisions'] = np.zeros(sphere0.n_cells, dtype=bool)`.  I get 

```txt
84
PolyData (0x7f49538ae760)
  N Cells:      1680
  N Points:     842
  X Bounds:     -4.993e-01, 4.993e-01
  Y Bounds:     -4.965e-01, 4.965e-01
  Z Bounds:     -5.000e-01, 5.000e-01
  N Arrays:     2

['ContactCells', 'Normals']
84
PolyData (0x7f495388a580)
  N Cells:      1680
  N Points:     842
  X Bounds:     -4.993e-01, 4.993e-01
  Y Bounds:     -4.965e-01, 4.965e-01
  Z Bounds:     -5.000e-01, 5.000e-01
  N Arrays:     3

['ContactCells', 'Normals']
```

Note that it _still_ says I have 3 arrays available, despite the original mesh only having 2.  I cannot see where the data from `generate_scalars` is available.  `sphere0` and `sphere1` look unchanged.

adeak commented:
Ha, yeah, even after `col.clear_arrays()` a phantom one seems to persist. Spooky!

darikg commented:
With `generate_scalars=True` the missing array is accessible through PlotterITK with an autogenerated name somehow  👻 

![Capture](https://user-images.githubusercontent.com/6875882/127566245-94bcd291-5163-44db-947e-d533fe733dab.PNG)

akaszynski commented:
I was going to gloss over this, but few things get past you guys, so thanks for pushing me to do the PRs right!

Discovered the magic behind the mystery data.  It's there as an active scalar and the array name is a null pointer:

```py
>>> import numpy as np
>>> import pyvista as pv
>>> sphere0 = pv.Sphere()
>>> sphere0['collisions'] = np.zeros(sphere0.n_cells, dtype=bool)
>>> sphere1 = pv.Sphere(radius=0.6, center=(-1, 0, 0))
>>> col, n_contacts = sphere0.collision(sphere1, generate_scalars=True)
>>> cdata = col.GetCellData()
>>> print(cdata)
vtkCellData (0xf935ad0)
  Debug: Off
  Modified Time: 1062652
  Reference Count: 2
  Registered Events: 
    Registered Observers:
      vtkObserver (0xf935c10)
        Event: 33
        EventName: ModifiedEvent
        Command: 0xf8e02e0
        Priority: 0
        Tag: 1
  Number Of Arrays: 1
  Array 0 name = nullptr
  Number Of Components: 4
  Number Of Tuples: 1680
  Copy Tuple Flags: ( 1 1 1 1 1 0 1 1 1 1 1 )
  Interpolate Flags: ( 1 1 1 1 1 0 0 1 1 1 1 )
  Pass Through Flags: ( 1 1 1 1 1 1 1 1 1 1 1 )
  Scalars: 
    Debug: Off
    Modified Time: 1062571
    Reference Count: 1
    Registered Events: (none)
    Name: (none)
    Data type: unsigned char
    Size: 6720
    MaxId: 6719
    NumberOfComponents: 4
    Information: 0
    Name: (none)
    Number Of Components: 4
    Number Of Tuples: 1680
    Size: 6720
    MaxId: 6719
    LookupTable: (none)
  Vectors: (none)
  Normals: (none)
  TCoords: (none)
  Tensors: (none)
  GlobalIds: (none)
  PedigreeIds: (none)
  EdgeFlag: (none)
  Tangents: (none)
  RationalWeights: (none)
  HigherOrderDegrees: (none)
```

Which means that we don't pick up on this array with:
```py
>>> col.cell_arrays.keys()
[]
```

However, it's there as an active scalar array!

```py
>>> col.cell_arrays.active_scalars
pyvista_ndarray([[255, 255, 255, 255],
                 [255, 255, 255, 255],
                 [255, 255, 255, 255],
                 ...,
                 [255, 255, 255, 255],
                 [255, 255, 255, 255],
                 [255, 255, 255, 255]], dtype=uint8)
```

Which means we can still plot it.

```py
col.plot(scalars=col.cell_arrays.active_scalars, rgb=True, cpos='zy', background='w')
```
![tmp](https://user-images.githubusercontent.com/11981631/127601410-aeb4e493-4e7f-4ae7-8488-2fd06f10d8c8.png)

akaszynski commented:
Since it turns out we don't support null array names, we're left with two choices:
1. Support null array names
2. Work around this limitation by naming the array that the filter generates.

I went with (2) since I'm not sure what fallout we'd get from converting a null array to a named array (like `'null'`.  Plus, it gives us a chance to name the array that comes off the filter.  This was done with:

```py
output.cell_arrays.GetAbstractArray(0).SetName('collision_rgba')
```

To get this to plot automatically as rgba rather than as double scalars, I've also added a feature that checks if the array name ends with rgb or rgba, it automatically enables `rgb=True`.  That way we can plot scalars simply with:

```py
import numpy as np
import pyvista
mesh_a = pyvista.Sphere(radius=0.5)
mesh_b = pyvista.Cube((0.5, 0.5, 0.5)).extract_cells([0, 2, 4])
collision, ncol = mesh_a.collision(mesh_b, cell_tolerance=1)
collision.plot()
```
![image](https://user-images.githubusercontent.com/11981631/127609150-aea3af34-e816-4dc7-85d5-0091bf86def3.png)

adeak commented:
I actually came across this last night, but I didn't realize that I struck gold:
```py
In [4]: col.get_array(0)
Out[4]: 
array([[255, 255, 255, 255],
       [255, 255, 255, 255],
       [255, 255, 255, 255],
       ...,
       [255, 255, 255, 255],
       [255, 255, 255, 255],
       [255, 255, 255, 255]], dtype=uint8)
```
This still worked after the other arrays were wiped. But this seems like a side-effect rather than null array name support: `pv.Sphere().get_array(0)` also returns the one scalar array (`'Normals'`).

MatthewFlamm commented on pyvista/core/filters/poly_data.py:
```diff
@@ -2601,3 +2601,155 @@ def strip(poly_data, join=False, max_length=1000, pass_cell_data=False,
+        ...                 show_edges=True)
+        >>> pl.show()
+
+        Alternatively, simply plot the colisions using the default
```
```suggestion
        Alternatively, simply plot the collisions using the default
```

akaszynski replied:
I'm terrible at spelling.  Thanks!

MatthewFlamm commented on pyvista/plotting/plotting.py:
```diff
@@ -1491,8 +1494,10 @@ def add_mesh(self, mesh, color=None, style=None, scalars=None,
+            those values as RGB(A) colors. ``rgba`` is also an
+            accepted alias for this.  Opacity (the A) is optional.  If
+            a scalars array ending with ``"_rgba"`` is passed, this is
+            automatically enabled.
```
Should this rather be something like "... the default becomes True"?  The user can still override it.

akaszynski replied:
True, that makes more sense.

MatthewFlamm reviewed (approved):
LGTM now.  I still don't quite get the usefulness of the scalars, but I think this is my misunderstanding as it now looks more like the vtk example.

akaszynski commented on pyvista/plotting/plotting.py:
```diff
@@ -1491,8 +1494,10 @@ def add_mesh(self, mesh, color=None, style=None, scalars=None,
-            alias for this.  Opacity (the A) is optional.
+            those values as RGB(A) colors. ``rgba`` is also an
+            accepted alias for this.  Opacity (the A) is optional.  If
+            a scalars array ending with ``"_rgba"`` is passed, this is
+            automatically enabled.
```
```suggestion
            a scalars array ending with ``"_rgba"`` is passed, the default
            becomes ``True``.  This can be overridden by setting this
            parameter to ``False``.
```
